### ISSUE TITLE
`er/expand-region` breaks in emacs 25 trunk build

### ISSUE THREAD

**Issue by kaushalmodi** [magnars/expand-region.el#160]
Hi,

I'd like to open this issue so that everyone using this package is aware of this.

As of [this][1] emacs build from trunk, `er/expand-region` does not work as expected. Here is an example; let's say that the **point is on the letter `c`** in the below text:

> abcde

On running `er/expand-region`, only `cde` get selected, not `abcde`! 

I was able to verify this by testing it out on a version which I had built from trunk on March 24 2015. That build does not show this problem.

A cursory glance at the [commits][2] between now and March 24th tells me this [this][3] commit could be causing this problem.

If you can recreate this issue and can confirm the proper cause of this issue, I can file a bug report.

[1]: http://git.savannah.gnu.org/cgit/emacs.git/commit/?id=2b828866c2df5ea558283e4c3f4c79a404918bea

[2]: http://git.savannah.gnu.org/cgit/emacs.git/log/

[3]: http://git.savannah.gnu.org/cgit/emacs.git/commit/?id=599ca626d760215b090012c69c749d391cfd6fbe

**Comment by PythonNut** [comment]
I can reproduce this.

**Comment by magnars** [comment]
I used to run integration tests against the emacs trunk, but it was always breaking. I've never seen any of those errors make it into a release, so I disabled the integration test. We'll have to fix the issue if it persists after a release.

**Comment by kaushalmodi** [comment]
@magnars This doesn't seem to be a bug but the way how the core behavior of `save-excursion` has changed. I believe that this problem is because the `save-excursion` function does NOT save the mark any more (it save only the point and buffer). Is there a way to modify `er--expand-region-1` to avert this change? I was able to narrow down the cause of this issue to [this code block](https://github.com/magnars/expand-region.el/blob/2832a0d04c5683c2ba3dc2b3c312eba2aa6b85b5/expand-region-core.el#L92-L102) but I am limited by my elisp skills to debug further.

**Comment by PythonNut** [comment]
Could we just manually save the mark with something like `(cl-letf (((mark))) ...)`?

**Comment by magnars** [comment]
Wow, I am stunned that they make such a large breaking change. I'm sure it's not only expand-region that breaks from this one. But yeah, we'll just change it if it makes to release.

**Comment by kaushalmodi** [comment]
@magnars Yeah I am unable to use the emacs trunk as my daily driver ever since that commit because I rely heavily on `expand-region` package. I am sure it would cause issues with other packages using `save-excursion` and `mark` but I am surprised I have heard anyone else complain about it (on reddit/emacs.stackexchange/G+ Community). 

@PythonNut Let me know if you want me to try out any solution to fix this.

**Comment by ffevotte** [comment]
FWIW, there's recently been a discussion about this on [emacs-devel](https://lists.gnu.org/archive/html/emacs-devel/2015-02/msg01254.html).

**Comment by magnars** [comment]
Thanks, I added my 2 cents there.

**Comment by jasonm23** [comment]
Quote: Stefan Monnier

> or I haven't looked in the right way. 

You think?!

There is quite literally, no reason to make this change and every reason to simply add *save-point*

I'm stunned this hasn't been squashed already.


Drew Adams nails it.

> +1 - Change requires a compelling argument in favor, not just
>     "Anyone here happen to see a compelling reason not to...?")

**Comment by jasonm23** [comment]
Apparently RMS also suggested **save-point**

**Comment by magnars** [comment]
Turns out the fix wasn't so small.

**Comment by kaushalmodi** [comment]
I haven't faced any problem with that fix.

**Comment by magnars** [comment]
No, but I have an extensive test suite that revealed several issues.

**Comment by kaushalmodi** [comment]
Cool! It will be good to put Stefan in the loop so that it can be decided if the save_excursion definition change needs to be reverted or if he can help modify expand_region to be compatible with the change.

**Comment by kaushalmodi** [comment]
Looks like we probably now have a [solution for this in emacs 25.0+][1]; use `save-mark-and-excursion` instead of `save-excursion`.

[1]: http://git.savannah.gnu.org/cgit/emacs.git/commit/?id=255a011f0ecf004b31c59945b10154b10fac3af1

**Comment by 11111000000** [comment]
kaushalmodi: this not works for me

**Comment by 11111000000** [comment]
So, how to fix this issue?

**Comment by kaushalmodi** [comment]
@11111000000 @magnars would need to edit the code to use the new `save-mark-and-excursion` function instead of `save-excursion`; the latest emacs version from its git master branch has this new function but it has to be used.

For now, I am using a [patched version of `er--expand-region-1`](https://github.com/kaushalmodi/.emacs.d/tree/a332a8f043c8457fb23e709eb68722db4de6d237/setup-files/setup-expand-region.el#L13) which works for all of my use cases.

**But** Magnar had [found some issue](https://github.com/magnars/expand-region.el/issues/160#issuecomment-92971668) using that patch in this package's test suite.

But, for me, the patch has worked fine for about a month now.

**Comment by cpitclaudel** [comment]
I can confirm that using `save-mark-and-excursion` in `er--expand-region-1` fixed it for me. Since I run two version of emacs in parallel I also added `(defalias 'save-mark-and-excursion 'save-excursion)` to my Emacs 24-specific startup file.

**Comment by magnars** [comment]
Sorry about the long delay on this one. I now use `save-mark-and-excursion`, and alias to for earlier Emacs versions.

**Comment by marcwebbie** [comment]
I am still having this issue on emacs 24.4

setting alias `(defalias 'save-mark-and-excursion 'save-excursion)` doesn't solve it for me:

```
er--expand-region-1: Invalid function: save-mark-and-excursion
```

**Comment by magnars** [comment]
Thanks for reporting, should be fixed in 69819ac.
